core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala

```diff
@@ -48,15 +49,22 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
   // in the presence of faults.
 
   /** Creates a mock [[BlockTransferService]] that returns data from the given map. */
-  private def createMockTransfer(data: Map[BlockId, ManagedBuffer]): BlockTransferService = {
+  private def createMockTransfer(
+      data: Map[BlockId, ManagedBuffer],
+      throwNettyOOM: Boolean = false): BlockTransferService = {
     val transfer = mock(classOf[BlockTransferService])
+    var hasThrowOOM = false
     when(transfer.fetchBlocks(any(), any(), any(), any(), any(), any())).thenAnswer(
       (invocation: InvocationOnMock) => {
         val blocks = invocation.getArguments()(3).asInstanceOf[Array[String]]
         val listener = invocation.getArguments()(4).asInstanceOf[BlockFetchingListener]
+        val OOMBlockIndex = new Random(System.currentTimeMillis()).nextInt(blocks.length)
 
-        for (blockId <- blocks) {
-          if (data.contains(BlockId(blockId))) {
+        for ((blockId, i) <- blocks.zipWithIndex) {
+          if (throwNettyOOM && !hasThrowOOM && i == OOMBlockIndex) {
+            hasThrowOOM = true
+            listener.onBlockFetchFailure(blockId, new TestNettyOutOfMemoryError())
```

Comment:
With two lines of code (+ 1 extra line import) you can remove the `TestNettyOutOfMemoryError.scala` file and keep the production code clean from the test implementation detail:

```suggestion
            val ctor = classOf[OutOfDirectMemoryError].getDeclaredConstructor(classOf[java.lang.String])
            ctor.setAccessible(true)
            listener.onBlockFetchFailure(blockId, ctor.newInstance("failed to allocate memory"))
```

The pattern (using `setAccessible`) is already used in Spark.

Reply:
This sounds good, I'll try it. Thanks!